### PR TITLE
Builder

### PR DESCRIPTION
**Pull request by Naandoo**

### CODE REVIEW

File: DesignPatternExamples/Assets/Design Patterns/Builder/Scripts/IAirshipBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+using System.Collections.Generic;
+
+namespace Builder
+{
+    public interface IAirshipBuilder
+    {
+        List<IAirshipModule> airshipModules { get; set; }
```
Comment (iceoffire) [inline DesignPatternExamples/Assets/Design Patterns/Builder/Scripts/IAirshipBuilder.cs:7]: why do you need a setter for this list?

Consider using `IList<IAirshipModule>`, and in the implementation uses a backingField which implements `List<IAirshipModule>`.


Consider doing the following in the implementation:
```
        private List<IAirshipModule> _airshipModules = new List<IAirshipModule>();
        public IList<IAirshipModule> airshipModules => _airshipModules.AsReadOnly();
```

The reasoning is that even without the Setter, lists can be manipulated by using the `List` class methods (e.g. `Add`). By returning an IList, and returning the `ReadOnly` version you prevent that from happening.

File: DesignPatternExamples/Assets/Design Patterns/Builder/Scripts/IAirshipBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+        void SetLeftWing(bool value);
+        void SetRightWing(bool value);
+        void SetLeftWeapon(bool value);
+        void SetRightWeapon(bool value);
+        void SetFrontalWeapon(bool value);
```
Comment (iceoffire) [inline DesignPatternExamples/Assets/Design Patterns/Builder/Scripts/IAirshipBuilder.cs:12]: Following the same principle of `value` parameter name should have context by the method name, consider changing it to the following:

```suggestion
        void SetLeftWingActive(bool value);
        void SetRightWingActive(bool value);
        void SetLeftWeaponActive(bool value);
        void SetRightWeaponActive(bool value);
        void SetFrontalWeaponActive(bool value);
```

File: DesignPatternExamples/Assets/Design Patterns/Builder/Scripts/IAirshipBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+        void GetAirship();
```
Comment (iceoffire) [inline DesignPatternExamples/Assets/Design Patterns/Builder/Scripts/IAirshipBuilder.cs:13]: It's strange that there's a `Get` method returning void, consider changing this name

File: DesignPatternExamples/Assets/Design Patterns/Builder/Scripts/AirshipBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            if (value) totalDamage += _airshipWeapon.attributeBonus;
+            else totalDamage -= _airshipWeapon.attributeBonus;
+            _airshipBuilderGUI.UpdateGUI(airshipModule);
+        }
+
+        private void UpdateSpeed(IAirshipModule airshipModule, bool value)
+        {
+            if (value) totalSpeed += _airshipWing.attributeBonus;
+            else totalSpeed -= _airshipWing.attributeBonus;
+            _airshipBuilderGUI.UpdateGUI(airshipModule);
```
[the remark this answers is not in the document]
Comment (Naandoo) [reply DesignPatternExamples/Assets/Design Patterns/Builder/Scripts/AirshipBuilder.cs:109]: I implemented a similar logic, by creating a Unity Event that's being invoked when the changes occur! I also subscribe manually the needed methods on the other classes.

I did the same for GetAirship method, that was using the same logic of calling update methods in other class.

File: DesignPatternExamples/Assets/Design Patterns/Builder/Scripts/Airship.cs [hunk 5]
```diff
@@ -0,0 +1,16 @@
+using System;
+using UnityEngine;
+
+namespace Builder
+{
+    public class Airship
```
[the remark this answers is not in the document]
Comment (Naandoo) [reply DesignPatternExamples/Assets/Design Patterns/Builder/Scripts/Airship.cs:6]: The initial idea was to hold all the information passed by the GUI interaction in an object and then use this object and its values in another system. But I didn't implement it, so I'll delete it.

File: DesignPatternExamples/Assets/Design Patterns/Builder/Scripts/IAirshipBuilder.cs (hunk 1 above)
Comment (Naandoo) [reply DesignPatternExamples/Assets/Design Patterns/Builder/Scripts/IAirshipBuilder.cs:7]: Incredible! Thanks!

File: DesignPatternExamples/Assets/Design Patterns/Builder/Scripts/AirshipBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        [SerializeField] private GameObject _rightWing;
+        [SerializeField] private AirshipBuilderGUI _airshipBuilderGUI;
+        public List<IAirshipModule> airshipModules { get; set; } = new List<IAirshipModule>();
+        [NonSerialized] public float totalSpeed = 10f;
+        [NonSerialized] public float totalDamage = 10f;
+        [NonSerialized] public bool hasLeftWing;
+        [NonSerialized] public bool hasRightWing;
+        private AirshipWeapon _airshipWeapon = new();
+        private AirshipWing _airshipWing = new();
+        private Airship _airship = new();
```
[the remark this answers is not in the document]
Comment (Naandoo) [reply DesignPatternExamples/Assets/Design Patterns/Builder/Scripts/AirshipBuilder.cs:23]: The _airshipWeapon and the _airshipWing, were created to represent the modules that make up the object being created.  I'm using it to calculate the amount of strength and speed that the airship has!

I'm using the bool hasRightWing and hasLeftWing to point if it's possible to use side weapons! Btw, I'm not sure if this is the best approach. 😅

The airship class indeed has no use right now.

File: DesignPatternExamples/Assets/Design Patterns/Builder/Scripts/AirshipBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+                case AirshipModuleType.Weapon:
+                    UpdateAttackDamage(airshipModule, value);
+                    break;
+                case AirshipModuleType.Wings:
+                    UpdateSpeed(airshipModule, value);
+                    break;
+            }
+        }
+
+        private void UpdateAttackDamage(IAirshipModule airshipModule, bool value)
```
[the remark this answers is not in the document]
Comment (Naandoo) [reply DesignPatternExamples/Assets/Design Patterns/Builder/Scripts/AirshipBuilder.cs:98]: You're right! In fact, it was a attention lack of mine!  The code wasn't directly, after the changes that you suggested, i don't even need to pass the bool values in the parameters of UpdateAttackDamage and UpdateAttackSpeed! But, i took the opportunity to see that "bool value" isn't clear, I'll make sure to be more descriptive in the next creations!

File: DesignPatternExamples/Assets/Design Patterns/Builder/Scripts/AirshipBuilderGUI.cs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+using DG.Tweening;
+
+namespace Builder
+{
+    public class AirshipBuilderGUI : MonoBehaviour
+    {
+        [SerializeField] private Image _attackFillBar;
+        [SerializeField] private Image _speedFillBar;
+        [SerializeField] private AirshipBuilder _airshipBuilder;
+        [SerializeField] Toggle _ToggleLeftWeapon;
```
[the remark this answers is not in the document]
Comment (Naandoo) [reply DesignPatternExamples/Assets/Design Patterns/Builder/Scripts/AirshipBuilderGUI.cs:12]: Lack of attention! I fixed!

**Pull request merged**